src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/EETypeNode.cs

```diff
@@ -94,25 +127,155 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly)
+
+                return false;
+            }
+        }
+
+        public override IEnumerable<CombinedDependencyListEntry> GetConditionalStaticDependencies(NodeFactory factory)
+        {
+            if (_type is MetadataType)
+            {
+                foreach (MethodDesc decl in VirtualFunctionResolution.EnumAllVirtualSlots((MetadataType)_type))
+                {
+                    MethodDesc impl = VirtualFunctionResolution.FindVirtualFunctionTargetMethodOnObjectType(decl, (MetadataType)_type);
+                    if (impl.OwningType == _type && !impl.IsAbstract)
+                    {
+                        yield return new DependencyNodeCore<NodeFactory>.CombinedDependencyListEntry(factory.MethodEntrypoint(impl), factory.VirtualMethodUse(decl), "Virtual method");
+                    }
+                }
+            }
+        }
+        
+        private void OutputComponentSize(ref ObjectDataBuilder objData)
+        {
             if (_type.IsArray)
             {
-                objData.EmitShort((short)((ArrayType)_type).ElementType.GetElementSize()); // m_ComponentSize
-                objData.EmitShort(0x4);                           // m_flags: IsArray(0x4)
+                objData.EmitShort((short)((ArrayType)_type).ElementType.GetElementSize());
             }
             else if (_type.IsString)
             {
-                objData.EmitShort(2); // m_ComponentSize
-                objData.EmitShort(0); // m_flags: 0
+                objData.EmitShort(2);
             }
             else
             {
-                objData.EmitShort(0); // m_ComponentSize
-                objData.EmitShort(0); // m_flags: 0
+                objData.EmitShort(0);
+            }
+        }
+        
+        private void OutputFlags(NodeFactory factory, ref ObjectDataBuilder objData)
+        {
+            // Todo: RelatedTypeViaIATFlag when we support cross-module EETypes
+            // Todo: GenericVarianceFlag when we support variance
+            // Todo: OptionalFieldsFlag
+
+            UInt16 flags = (UInt16)EETypeFlags.CanonicalEEType;
+            
+            if (_type.IsArray || _type.IsPointer)
+            {
+                flags = (UInt16)EETypeFlags.ParameterizedEEType;
+            }
+            
+            if (_type.IsValueType)
+            {
+                flags |= (UInt16)EETypeFlags.ValueTypeFlag;
+            }
+
+            if (_type.HasFinalizer)
+            {
+                flags |= (UInt16)EETypeFlags.HasFinalizerFlag;
+            }
+            
+            if (_type is MetadataType && ((MetadataType)_type).ContainsPointers)
+            {
+                flags |= (UInt16)EETypeFlags.HasPointersFlag;
+            }
+            else if (_type.IsArray)
+            {
+                ArrayType arrayType = _type as ArrayType;
+                if ((arrayType.ElementType.IsValueType && arrayType.ElementType is MetadataType && ((MetadataType)arrayType.ElementType).ContainsPointers) ||
```

Comment:
If IsValueType is true, you can always safely cast to DefType. And DefType has ContainsPointers.
(i.e. skip the `is` check.)